src/main/java/com/google/googleidentity/oauth2/risc/RiscHandler.java

```diff
@@ -0,0 +1,172 @@
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+package com.google.googleidentity.oauth2.risc;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.hash.Hashing;
+import com.google.googleidentity.oauth2.client.ClientDetails;
+import com.google.googleidentity.oauth2.client.ClientDetailsService;
+import com.google.googleidentity.oauth2.jwt.JwkStore;
+import com.google.googleidentity.oauth2.token.OAuth2AccessToken;
+import com.google.googleidentity.oauth2.token.OAuth2RefreshToken;
+import com.google.googleidentity.oauth2.token.OAuth2TokenService;
+import com.google.googleidentity.oauth2.util.OAuth2Constants.TokenTypes;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jose.util.StandardCharset;
+import io.jsonwebtoken.Jwts;
+import java.io.IOException;
+import java.security.Key;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.entity.EntityBuilder;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+
+/** Revoke tokens between a user and a client. Send risc if the client support it. */
+@Singleton
+public class RiscHandler {
+  private static final Logger log = Logger.getLogger("RiscHandler");
+  private static int MAX_RETRY_COUNT = 4;
+  private static Duration RETRY_INTERVAL_TIME = Duration.ofMinutes(10);
+  private static String WEB_URL =
+      System.getenv("WEB_URL") == null ? "localhost:8080" : System.getenv("WEB_URL");
+  private final ClientDetailsService clientDetailsService;
+  private final JwkStore jwkStore;
+  private long jtiValue = 0l;
+
+  @Inject
+  public RiscHandler(JwkStore jwkStore, ClientDetailsService clientDetailsService) {
+    this.jwkStore = jwkStore;
+    this.clientDetailsService = clientDetailsService;
+  }
+
+  public void SendRisc(
+      List<OAuth2AccessToken> accessTokenList, Optional<OAuth2RefreshToken> refreshToken) {
+
+    for (OAuth2AccessToken token : accessTokenList) {
+      Thread thread = new sendEventThread(token);
+      thread.start();
+    }
+
+    if (refreshToken.isPresent()) {
+      Thread thread = new sendEventThread(refreshToken.get());
+      thread.start();
+    }
+  }
+
+  private synchronized long getJtiValue() {
+    return jtiValue++;
```

Comment:
This can hardly ensure the uniqueness of the event id. Because it will reset every time restart the server. I think we can use hash(client id + username + timestamp + random) to generate the id.